File: scanner/scanner.js
// TODO: not tested
let _buffer = '';

function _feed_buffer() {
    while (_buffer.trim().length === 0) {
        _buffer = prompt();
    }
}

function _read_number(number_type) {
    _feed_buffer();

    const match = _buffer.match(/\s*(-?\d+([.]\d+)?)*/);
    if (match) {
        _buffer = _buffer.slice(match[0].length);
        return number_type(match[1]);
    } else {
        throw new Error('Expected integer');
    }
}

function read_int() {
    return _read_number(parseInt);
}

function read_float() {
    return _read_number(parseFloat);
}

function read_line() {
    _feed_buffer();
    return _buffer;
}

function read_word() {
    _feed_buffer();

    const match = _buffer.match(/\s*([^\s]+)*/);
    if (match) {
        _buffer = _buffer.slice(match[0].length);
        return match[1];
    } else {
        throw new Error();
    }
}

File: scanner/scanner.py
import re

_buffer = ''

def _feed_buffer():
    global _buffer
    while len(_buffer.strip()) == 0:
        _buffer = input()

def read_number(number_type = None):
    global _buffer
    _feed_buffer()

    if match := re.match(r'\s*(-?\d+([.]\d+)?)', _buffer):            
        _buffer = _buffer[match.end():]
        if number_type is None:
            if '.' in match.group(1):
                return float(match.group(1))
            else:
                return int(match.group(1))
        else:
            return number_type(match.group(1))
    else:
        raise ValueError('Expected integer')

def read_int():
    return read_number(int)
def read_float():
    return read_number(float)

def read_line():
    global _buffer
    _feed_buffer()
    return _buffer

def read_word():
    global _buffer
    _feed_buffer()
    if match := re.match(r'\s*([^\s]+)', _buffer):            
        _buffer = _buffer[match.end():]
        return match.group(1)
    else:
        raise ValueError()